File: 3.cpp
#include <array>
#include <string>

class Solution {
public:
	int lengthOfLongestSubstring(const std::string &s) {
		int longest = 0;
		std::array<bool, 256> alreadySeen{false};
		int left = 0;
		int right = 0;
		while(right < s.length()) {
			while(right < s.length() && !alreadySeen[s[right]]) {
				alreadySeen[s[right]] = true;
				right++;
			}
			int len = right - left;
			if(len > longest) {
				longest = len;
			}
			alreadySeen[s[left]] = false;
			left++;
		}
		return longest;
	}
};

int main() {
	Solution s;
	const auto result = s.lengthOfLongestSubstring("abcabcbb");
	return 0;
}

File: 3.ts
function lengthOfLongestSubstring(s: string): number {
    if(s.length === 0) {
        return 0;
    }
    const alreadySeen: boolean[] = new Array(128).fill(false);
    var left = 0;
    var right = 0;
    var bestLen = 0;
    while(right < s.length) {
        const cR = s.charCodeAt(right);
        if(!alreadySeen[cR]) {
            alreadySeen[cR] = true;
            right++;
            bestLen = Math.max(bestLen, right - left);
        } else {
            const cL = s.charCodeAt(left);
            alreadySeen[cL] = false;
            left++;
        }
    }
    return bestLen;

}
